nexus/src/app/sagas/instance_create.rs

```diff
@@ -299,11 +311,255 @@ impl NexusSaga for SagaInstanceCreate {
+
+    let instance_id = db_instance.id();
+    let sled_uuid = db_instance.runtime_state.sled_id;
+
+    let (.., sled) = LookupPath::new(&osagactx.nexus().opctx_alloc, &datastore)
+        .sled_id(sled_uuid)
+        .fetch()
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    let sled_ip_address = sled.address();
+
+    debug!(log, "fetching network interfaces");
+
+    let network_interface = match datastore
+        .derive_guest_network_interface_info(&opctx, &authz_instance)
+        .await
+        .map_err(ActionError::action_failed)?
+        .into_iter()
+        .find(|interface| interface.primary)
+    {
+        Some(interface) => interface,
+        // Return early if instance does not have a primary network
+        // interface
+        None => return Ok(()),
+    };
+
+    let mac_address =
+        macaddr::MacAddr6::from_str(&network_interface.mac.to_string())
+            .map_err(|e| {
+                ActionError::action_failed(Error::internal_error(&format!(
+                    "failed to convert mac address: {e}"
+                )))
+            })?;
+
+    let vni: u32 = network_interface.vni.into();
+
+    debug!(log, "fetching external ip addresses");
+
+    let target_ip = &datastore
+        .instance_lookup_external_ips(&opctx, instance_id)
+        .await
+        .map_err(ActionError::action_failed)?
+        .get(which)
+        .ok_or_else(|| {
+            ActionError::action_failed(Error::internal_error(&format!(
+                "failed to find external ip address at index: {which}"
+            )))
+        })?
+        .to_owned();
+
+    debug!(log, "checking for existing nat mapping for {target_ip:#?}");
+
+    // TODO: currently if we have this environment variable set, we want to
+    // bypass all calls to DPD. This is mainly to facilitate some tests where
+    // we don't have dpd running. In the future we should probably have these
+    // testing environments running dpd-stub so that the full path can be tested.
+    if let Ok(_) = std::env::var("SKIP_ASIC_CONFIG") {
+        return Ok(());
+    };
+
+    let existing_nat = match target_ip.ip {
+        ipnetwork::IpNetwork::V4(network) => {
+            dpd_client.nat_ipv4_get(&network.ip(), *target_ip.first_port).await
+        }
+        ipnetwork::IpNetwork::V6(network) => {
+            dpd_client.nat_ipv6_get(&network.ip(), *target_ip.first_port).await
+        }
+    };
+
+    match existing_nat {
+        Ok(_) => {
+            // nat entry already exists, do nothing
+            return Ok(());
+        }
+        Err(e) => {
+            if e.status() == Some(http::StatusCode::NOT_FOUND) {
+                debug!(log, "no nat entry found for: {target_ip:#?}");
+            } else {
+                return Err(ActionError::action_failed(Error::internal_error(
+                    &format!("failed to query dpd: {e}"),
+                )));
+            }
+        }
+    }
+
```

Comment:
I'm trying to remember the last conversation we had about this, but I thought we decided that the error-handling here would be a bit different. We'd instead try to create the mapping, and ignore an "existing mapping" error, rather than checking if it exists and then creating it.

Is that still worth doing here?

Reply:
I _think_ the reason we changed it in the various `delete` methods but not here was because we could match on `Some(http::StatusCode::NOT_FOUND)` in the deletion attempts but we didn't get something similar back during creation attempts?

Reply:
Ugh, right. See https://github.com/oxidecomputer/dendrite/issues/152